RIOT-OS/RIOT
# Simple duty cycling 802.15.4 MAC protocol

I have developed a (not so) simple (anymore) duty cycling MAC protocol focused for IEEE 802.15.4 transceiver. It has the working title "LwMAC" which stands for "lightweight MAC", I really didn't have a better idea. This is still work-in-progress but has reached a point where maybe someone else might want to try it or comment on it's future development.

In the current state it behaves like the ~~`LPEA`~~ `LPEAS` protocol described in [this paper](http://ieeexplore.ieee.org/xpls/abs_all.jsp?arnumber=4539479) (sorry, there doesn't seem to be a free source). This is quite similiar to [ContikiMAC](http://dunkels.com/adam/dunkels11contikimac.pdf), but doesn't provide compatibility (yet). ~~Later it's planned to add implicit synchronization with neighbor nodes, i.e. extent to `LPEAS`.~~ Done

I created a quick [overview](http://docdro.id/sbNzH7D) of the overall structure. Maybe this helps to better understand all the moving parts. Yet there is no overview over the RX or TX state machines.

It still has bugs and TODOs, so I want to use this PR as my TODO list.

Currently this depends on ~~#3138~~ ~~#3619~~ ~~#3729~~ ~~#4078~~ ~~#3867~~ ~~#4081~~ ~~#2309~~  
Incorporates, but not critical: ~~#3993~~ ~~#3868~~ ~~#3887~~ ~~#4015~~ #4058  ~~#3954~~ 
# TODOs
- ~~use gnrc pktqueue~~
- ~~implement broadcast~~
- ~~verify duty cycling~~
- ~~implement implicit sync~~
- use xtimer instead of RTT peripheral when low power timer source is implemented (#3990 et al.)
- update documentation, make clear what's supposed to be changed by the user
- rework error logging
- integrate properly and implement selection of mac layer
- review and address inline TODOs
- major cleanup
# How to test

Currently you should use the `samr21-xpro` board to test this, because certain features (see depending PRs) are only available on that platform. And it's the only I can test :-)

Flash to board with the at86rf2xx driver test:

``` bash
$ cd tests/driver_at86rf2xx
$ BOARD=samr21-xpro make flash term
```

Then manually send packet from one board to the other via interactive shell(~~**note:** baudrate is 921600~~) 

```
txtsnd 4 xx:xx Helloxx
```

or even as broadcast:

```
txtsnd 4 bcast aReallyBroadCast
```

You can use a packet sniffer to verify the communication or just check if packets arrive at the destination by flashing the same test.
